// permissionless@0.0.6/_cjs/actions/public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccountNonce = exports.getSenderAddress = exports.InvalidEntryPointError = void 0;
const viem_1 = require("viem");
class InvalidEntryPointError extends viem_1.BaseError {
    constructor({ cause, entryPoint } = {}) {
        super(`The entry point address (\`entryPoint\`${entryPoint ? ` = ${entryPoint}` : ""}) is not a valid entry point. getSenderAddress did not revert with a SenderAddressResult error.`, {
            cause
        });
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "InvalidEntryPointError"
        });
    }
}
exports.InvalidEntryPointError = InvalidEntryPointError;
const getSenderAddress = async (publicClient, { initCode, entryPoint }) => {
    try {
        await publicClient.simulateContract({
            address: entryPoint,
            abi: [
                {
                    inputs: [
                        {
                            internalType: "address",
                            name: "sender",
                            type: "address"
                        }
                    ],
                    name: "SenderAddressResult",
                    type: "error"
                },
                {
                    inputs: [
                        {
                            internalType: "bytes",
                            name: "initCode",
                            type: "bytes"
                        }
                    ],
                    name: "getSenderAddress",
                    outputs: [],
                    stateMutability: "nonpayable",
                    type: "function"
                }
            ],
            functionName: "getSenderAddress",
            args: [initCode]
        });
    }
    catch (e) {
        const err = e;
        if (err.cause.name === "ContractFunctionRevertedError") {
            const revertError = err.cause;
            const errorName = revertError.data?.errorName ?? "";
            if (errorName === "SenderAddressResult" && revertError.data?.args && revertError.data?.args[0]) {
                return revertError.data?.args[0];
            }
        }
        throw e;
    }
    throw new InvalidEntryPointError({ entryPoint });
};
exports.getSenderAddress = getSenderAddress;
const getAccountNonce = async (publicClient, { address, entryPoint, key = 0n }) => {
    return await publicClient.readContract({
        address: entryPoint,
        abi: [
            {
                inputs: [
                    {
                        name: "sender",
                        type: "address"
                    },
                    {
                        name: "key",
                        type: "uint192"
                    }
                ],
                name: "getNonce",
                outputs: [
                    {
                        name: "nonce",
                        type: "uint256"
                    }
                ],
                stateMutability: "view",
                type: "function"
            }
        ],
        functionName: "getNonce",
        args: [address, key]
    });
};
exports.getAccountNonce = getAccountNonce;
